File: frontend/lib/tools/pinterestFit/engine.ts
import {
    BLOCKER_REASON_PRIORITY,
    MAX_SCORE,
    NOT_RIGHT_NOW_FALLBACK_PRIORITY,
    OUTCOME_SCORE_BANDS,
    POSITIVE_REASON_PRIORITY,
    ROLE_PRIORITY_ORDER,
} from "./config";
import { CTA_BY_OUTCOME, REASON_COPY_BY_KEY, RESULT_COPY_BY_OUTCOME, ROLE_COPY_BY_KEY } from "./copy";
import { PINTEREST_FIT_QUESTIONS_BY_ID } from "./questions";
import type {
    AdsOpennessValue,
    AssessmentAnswerValue,
    AssessmentAnswers,
    AssessmentQuestionOption,
    AssessmentResult,
    AssessmentSignalKey,
    AssetsValue,
    CategoryFitValue,
    GoalFitValue,
    GoalType,
    GuardrailKey,
    OutcomeKey,
    OfferProvenValue,
    ReasonKey,
    ResultReasons,
    ResultReasonKeys,
    RoleKey,
    StoredAssessmentAnswers,
    SupportReadinessValue,
    WebsiteReadinessValue,
} from "./types";

type SignalEvaluation = Readonly<{
    signal: AssessmentSignalKey;
    score: number;
    reasonKey: ReasonKey;
    isPositive: boolean;
    isBlocker: boolean;
    isModerate: boolean;
}>;

type SignalEvaluationMap = Readonly<Record<AssessmentSignalKey, SignalEvaluation>>;

const FOUNDATION_SIGNALS = ["offer", "assets", "website"] as const satisfies readonly AssessmentSignalKey[];
const READINESS_INTENT_SIGNALS = [
    "support_readiness",
    "ads_openness",
    "goal",
] as const satisfies readonly AssessmentSignalKey[];

function getSelectedOption<TValue extends AssessmentAnswerValue>(
    questionId: keyof typeof PINTEREST_FIT_QUESTIONS_BY_ID,
    value: TValue,
): AssessmentQuestionOption<TValue> {
    const option = PINTEREST_FIT_QUESTIONS_BY_ID[questionId].options.find((candidate) => candidate.value === value);

    if (!option) {
        throw new Error(`Unknown Pinterest Fit answer value "${value}" for ${questionId}.`);
    }

    return option as AssessmentQuestionOption<TValue>;
}

export function normalizePinterestFitAnswers(answers: AssessmentAnswers): StoredAssessmentAnswers {
    const q1 = getSelectedOption<CategoryFitValue>("q1", answers.q1);
    const q2 = getSelectedOption<OfferProvenValue>("q2", answers.q2);
    const q3 = getSelectedOption<AssetsValue>("q3", answers.q3);
    const q4 = getSelectedOption<WebsiteReadinessValue>("q4", answers.q4);
    const q5 = getSelectedOption<GoalFitValue>("q5", answers.q5);
    const q6 = getSelectedOption<SupportReadinessValue>("q6", answers.q6);
    const q7 = getSelectedOption<AdsOpennessValue>("q7", answers.q7);

    if (!q5.goalType) {
        throw new Error(`Pinterest Fit goal answer "${answers.q5}" is missing a goalType mapping.`);
    }

    return {
        q1_category_fit: q1.score,
        q2_offer_proven: q2.score,
        q3_assets: q3.score,
        q4_website: q4.score,
        q5_goal_fit: q5.score,
        q5_goal_type: q5.goalType,
        q6_support_readiness: q6.score,
        q7_ads_openness: q7.score,
    };
}

function calculateTotalScore(normalizedAnswers: StoredAssessmentAnswers): number {
    return (
        normalizedAnswers.q1_category_fit +
        normalizedAnswers.q2_offer_proven +
        normalizedAnswers.q3_assets +
        normalizedAnswers.q4_website +
        normalizedAnswers.q5_goal_fit +
        normalizedAnswers.q6_support_readiness +
        normalizedAnswers.q7_ads_openness
    );
}

function resolveBaseOutcome(score: number): OutcomeKey {
    const matchingBand = OUTCOME_SCORE_BANDS.find((band) => score >= band.minScore && score <= band.maxScore);

    if (!matchingBand) {
        throw new Error(`Pinterest Fit score "${score}" fell outside configured outcome bands.`);
    }

    return matchingBand.outcome;
}

function applyGuardrails(normalizedAnswers: StoredAssessmentAnswers, baseOutcome: OutcomeKey): {
    finalOutcome: OutcomeKey;
    triggeredGuardrails: readonly GuardrailKey[];
} {
    const triggeredGuardrails: GuardrailKey[] = [];
    let finalOutcome = baseOutcome;

    const guardrailA =
        normalizedAnswers.q5_goal_type === "sales" &&
        normalizedAnswers.q6_support_readiness <= 1 &&
        normalizedAnswers.q7_ads_openness <= 1;

    if (guardrailA) {
        triggeredGuardrails.push("guardrail_a");
        if (finalOutcome === "strong_fit") {
            finalOutcome = "possible_fit";
        }
    }

    const guardrailB = normalizedAnswers.q3_assets === 0 && normalizedAnswers.q4_website === 0;
    if (guardrailB) {
        triggeredGuardrails.push("guardrail_b");
        finalOutcome = "not_right_now";
    }

    const guardrailC = normalizedAnswers.q1_category_fit === 1 && normalizedAnswers.q2_offer_proven <= 1;
    if (guardrailC) {
        triggeredGuardrails.push("guardrail_c");
        finalOutcome = "not_right_now";
    }

    return {
        finalOutcome,
        triggeredGuardrails,
    };
}

function getSignalScore(signal: AssessmentSignalKey, normalizedAnswers: StoredAssessmentAnswers): number {
    switch (signal) {
        case "category":
            return normalizedAnswers.q1_category_fit;
        case "offer":
            return normalizedAnswers.q2_offer_proven;
        case "assets":
            return normalizedAnswers.q3_assets;
        case "website":
            return normalizedAnswers.q4_website;
        case "goal":
            return normalizedAnswers.q5_goal_fit;
        case "support_readiness":
            return normalizedAnswers.q6_support_readiness;
        case "ads_openness":
            return normalizedAnswers.q7_ads_openness;
    }
}

function getSignalReasonKey(signal: AssessmentSignalKey, normalizedAnswers: StoredAssessmentAnswers): ReasonKey {
    switch (signal) {
        case "category":
            switch (normalizedAnswers.q1_category_fit) {
                case 4:
                    return "reason_category_strong";
                case 3:
                    return "reason_category_good";
                case 2:
                    return "reason_category_maybe";
                case 1:
                    return "reason_category_weak";
                default:
                    throw new Error(`Unsupported Pinterest Fit category score "${normalizedAnswers.q1_category_fit}".`);
            }
        case "offer":
            switch (normalizedAnswers.q2_offer_proven) {
                case 4:
                    return "reason_offer_proven";
                case 3:
                    return "reason_offer_some_traction";
                case 1:
                    return "reason_offer_early";
                case 0:
                    return "reason_offer_unproven";
                default:
                    throw new Error(`Unsupported Pinterest Fit offer score "${normalizedAnswers.q2_offer_proven}".`);
            }
        case "assets":
            switch (normalizedAnswers.q3_assets) {
                case 4:
                    return "reason_assets_strong";
                case 3:
                    return "reason_assets_decent";
                case 1:
                    return "reason_assets_limited";
                case 0:
                    return "reason_assets_weak";
                default:
                    throw new Error(`Unsupported Pinterest Fit assets score "${normalizedAnswers.q3_assets}".`);
            }
        case "website":
            switch (normalizedAnswers.q4_website) {
                case 4:
                    return "reason_site_ready";
                case 3:
                    return "reason_site_solid";
                case 1:
                    return "reason_site_friction";
                case 0:
                    return "reason_site_not_ready";
                default:
                    throw new Error(`Unsupported Pinterest Fit website score "${normalizedAnswers.q4_website}".`);
            }
        case "goal":
            return getGoalReasonKey(normalizedAnswers.q5_goal_type);
        case "support_readiness":
            switch (normalizedAnswers.q6_support_readiness) {
                case 3:
                    return "reason_support_ready";
                case 2:
                    return "reason_support_open";
                case 1:
                    return "reason_support_cautious";
                case 0:
                    return "reason_support_not_committed";
                default:
                    throw new Error(
                        `Unsupported Pinterest Fit support readiness score "${normalizedAnswers.q6_support_readiness}".`,
                    );
            }
        case "ads_openness":
            switch (normalizedAnswers.q7_ads_openness) {
                case 3:
                    return "reason_ads_open";
                case 2:
                    return "reason_ads_later";
                case 1:
                    return "reason_ads_unsure";
                case 0:
                    return "reason_ads_not_open";
                default:
                    throw new Error(`Unsupported Pinterest Fit ads openness score "${normalizedAnswers.q7_ads_openness}".`);
            }
    }
}

function getGoalReasonKey(goalType: GoalType): ReasonKey {
    switch (goalType) {
        case "discovery":
            return "reason_goal_discovery";
        case "traffic":
            return "reason_goal_traffic";
        case "launches":
            return "reason_goal_launches";
        case "retargeting":
            return "reason_goal_retargeting";
        case "sales":
            return "reason_goal_sales_caution";
    }
}

function isPositiveSignal(signal: AssessmentSignalKey, normalizedAnswers: StoredAssessmentAnswers): boolean {
    switch (signal) {
        case "category":
            return normalizedAnswers.q1_category_fit >= 3;
        case "offer":
            return normalizedAnswers.q2_offer_proven >= 3;
        case "assets":
            return normalizedAnswers.q3_assets >= 3;
        case "website":
            return normalizedAnswers.q4_website >= 3;
        case "goal":
            return normalizedAnswers.q5_goal_type !== "sales";
        case "support_readiness":
            return normalizedAnswers.q6_support_readiness >= 2;
        case "ads_openness":
            return normalizedAnswers.q7_ads_openness >= 2;
    }
}

function isBlockerSignal(signal: AssessmentSignalKey, normalizedAnswers: StoredAssessmentAnswers): boolean {
    switch (signal) {
        case "category":
            return normalizedAnswers.q1_category_fit <= 2;
        case "offer":
            return normalizedAnswers.q2_offer_proven <= 1;
        case "assets":
            return normalizedAnswers.q3_assets <= 1;
        case "website":
            return normalizedAnswers.q4_website <= 1;
        case "goal":
            return normalizedAnswers.q5_goal_type === "sales";
        case "support_readiness":
            return normalizedAnswers.q6_support_readiness <= 1;
        case "ads_openness":
            return normalizedAnswers.q7_ads_openness <= 1;
    }
}

function buildSignalEvaluations(normalizedAnswers: StoredAssessmentAnswers): SignalEvaluationMap {
    return POSITIVE_REASON_PRIORITY.reduce<Record<AssessmentSignalKey, SignalEvaluation>>((accumulator, signal) => {
        const isPositive = isPositiveSignal(signal, normalizedAnswers);
        const isBlocker = isBlockerSignal(signal, normalizedAnswers);

        accumulator[signal] = {
            signal,
            score: getSignalScore(signal, normalizedAnswers),
            reasonKey: getSignalReasonKey(signal, normalizedAnswers),
            isPositive,
            isBlocker,
            isModerate: !isPositive && !isBlocker,
        };

        return accumulator;
    }, {} as Record<AssessmentSignalKey, SignalEvaluation>);
}

function selectHighestScoringSignal(
    evaluations: SignalEvaluationMap,
    priorityOrder: readonly AssessmentSignalKey[],
): AssessmentSignalKey {
    return priorityOrder.reduce<AssessmentSignalKey>((bestSignal, signal) => {
        if (evaluations[signal].score > evaluations[bestSignal].score) {
            return signal;
        }

        return bestSignal;
    }, priorityOrder[0]);
}

function selectStrongFitReasonSignals(evaluations: SignalEvaluationMap): readonly AssessmentSignalKey[] {
    return [
        "category",
        selectHighestScoringSignal(evaluations, FOUNDATION_SIGNALS),
        selectHighestScoringSignal(evaluations, READINESS_INTENT_SIGNALS),
    ] as const;
}

function selectPossibleFitReasonSignals(evaluations: SignalEvaluationMap): readonly AssessmentSignalKey[] {
    const selectedSignals: AssessmentSignalKey[] = [];
    const positiveSignals = POSITIVE_REASON_PRIORITY.filter((signal) => evaluations[signal].isPositive);
    const moderateSignals = POSITIVE_REASON_PRIORITY.filter((signal) => evaluations[signal].isModerate);

    for (const signal of positiveSignals) {
        if (selectedSignals.length === 2) {
            break;
        }

        selectedSignals.push(signal);
    }

    for (const signal of moderateSignals) {
        if (selectedSignals.length === 2) {
            break;
        }

        if (!selectedSignals.includes(signal)) {
            selectedSignals.push(signal);
        }
    }

    for (const signal of POSITIVE_REASON_PRIORITY) {
        if (selectedSignals.length === 2) {
            break;
        }

        if (!selectedSignals.includes(signal)) {
            selectedSignals.push(signal);
        }
    }

    const blockerSignal = BLOCKER_REASON_PRIORITY.find(
        (signal) => evaluations[signal].isBlocker && !selectedSignals.includes(signal),
    );

    if (blockerSignal) {
        selectedSignals.push(blockerSignal);
    } else {
        const fallbackSignal = BLOCKER_REASON_PRIORITY.find((signal) => !selectedSignals.includes(signal));

        if (!fallbackSignal) {
            throw new Error("Pinterest Fit possible-fit reason selection could not resolve a third reason.");
        }

        selectedSignals.push(fallbackSignal);
    }

    return selectedSignals as readonly AssessmentSignalKey[];
}

function selectNotRightNowReasonSignals(evaluations: SignalEvaluationMap): readonly AssessmentSignalKey[] {
    const selectedSignals = BLOCKER_REASON_PRIORITY.filter((signal) => evaluations[signal].isBlocker).slice(0, 3);

    if (selectedSignals.length < 3) {
        const fallbackIndexBySignal = NOT_RIGHT_NOW_FALLBACK_PRIORITY.reduce<Record<AssessmentSignalKey, number>>(
            (accumulator, signal, index) => {
                accumulator[signal] = index;
                return accumulator;
            },
            {} as Record<AssessmentSignalKey, number>,
        );

        const remainingSignals = POSITIVE_REASON_PRIORITY.filter((signal) => !selectedSignals.includes(signal)).sort(
            (left, right) => {
                const scoreDifference = evaluations[left].score - evaluations[right].score;

                if (scoreDifference !== 0) {
                    return scoreDifference;
                }

                return fallbackIndexBySignal[left] - fallbackIndexBySignal[right];
            },
        );

        for (const signal of remainingSignals) {
            if (selectedSignals.length === 3) {
                break;
            }

            selectedSignals.push(signal);
        }
    }

    if (selectedSignals.length !== 3) {
        throw new Error("Pinterest Fit not-right-now reason selection did not resolve exactly three reasons.");
    }

    return selectedSignals as readonly AssessmentSignalKey[];
}

function selectReasonSignals(
    finalOutcome: OutcomeKey,
    evaluations: SignalEvaluationMap,
): readonly AssessmentSignalKey[] {
    switch (finalOutcome) {
        case "strong_fit":
            return selectStrongFitReasonSignals(evaluations);
        case "possible_fit":
            return selectPossibleFitReasonSignals(evaluations);
        case "not_right_now":
            return selectNotRightNowReasonSignals(evaluations);
    }
}

function mapReasonKeys(
    selectedSignals: readonly AssessmentSignalKey[],
    evaluations: SignalEvaluationMap,
): ResultReasonKeys {
    if (selectedSignals.length !== 3) {
        throw new Error("Pinterest Fit reason selection must resolve exactly three signals.");
    }

    return [
        evaluations[selectedSignals[0]].reasonKey,
        evaluations[selectedSignals[1]].reasonKey,
        evaluations[selectedSignals[2]].reasonKey,
    ];
}

function mapReasonCopy(reasonKeys: ResultReasonKeys): ResultReasons {
    return [
        REASON_COPY_BY_KEY[reasonKeys[0]],
        REASON_COPY_BY_KEY[reasonKeys[1]],
        REASON_COPY_BY_KEY[reasonKeys[2]],
    ];
}

function matchesRole(
    role: RoleKey,
    normalizedAnswers: StoredAssessmentAnswers,
    finalOutcome: OutcomeKey,
): boolean {
    switch (role) {
        case "not_priority_yet":
            return finalOutcome === "not_right_now";
        case "sales_with_ads_support":
            return (
                normalizedAnswers.q5_goal_type === "sales" &&
                normalizedAnswers.q2_offer_proven >= 3 &&
                normalizedAnswers.q4_website >= 3 &&
                normalizedAnswers.q7_ads_openness >= 2
            );
        case "warm_audience_support":
            return normalizedAnswers.q5_goal_type === "retargeting" && finalOutcome !== "not_right_now";
        case "discovery_traffic":
            return (
                normalizedAnswers.q1_category_fit >= 3 &&
                normalizedAnswers.q3_assets >= 3 &&
                normalizedAnswers.q4_website >= 3 &&
                (normalizedAnswers.q5_goal_type === "discovery" || normalizedAnswers.q5_goal_type === "traffic")
            );
        case "selective_test_channel":
            return (
                normalizedAnswers.q1_category_fit <= 2 &&
                normalizedAnswers.q2_offer_proven >= 3 &&
                normalizedAnswers.q3_assets >= 3 &&
                normalizedAnswers.q4_website >= 3 &&
                (normalizedAnswers.q5_goal_type === "discovery" ||
                    normalizedAnswers.q5_goal_type === "traffic" ||
                    normalizedAnswers.q5_goal_type === "launches") &&
                finalOutcome !== "not_right_now"
            );
        case "organic_first_ads_later":
            return (
                normalizedAnswers.q1_category_fit >= 3 &&
                normalizedAnswers.q3_assets >= 1 &&
                normalizedAnswers.q4_website >= 1 &&
                (normalizedAnswers.q5_goal_type === "discovery" ||
                    normalizedAnswers.q5_goal_type === "traffic" ||
                    normalizedAnswers.q5_goal_type === "launches")
            );
        case "foundation_first":
            return finalOutcome !== "not_right_now";
    }
}

function resolveRole(normalizedAnswers: StoredAssessmentAnswers, finalOutcome: OutcomeKey): RoleKey {
    const matchedRole = ROLE_PRIORITY_ORDER.find((role) => matchesRole(role, normalizedAnswers, finalOutcome));

    if (!matchedRole) {
        throw new Error("Pinterest Fit role resolution failed to match a role.");
    }

    return matchedRole;
}

export function scorePinterestFitAssessment(answers: AssessmentAnswers): AssessmentResult {
    const normalizedAnswers = normalizePinterestFitAnswers(answers);
    const score = calculateTotalScore(normalizedAnswers);
    const baseOutcome = resolveBaseOutcome(score);
    const { finalOutcome, triggeredGuardrails } = applyGuardrails(normalizedAnswers, baseOutcome);
    const evaluations = buildSignalEvaluations(normalizedAnswers);
    const selectedReasonSignals = selectReasonSignals(finalOutcome, evaluations);
    const reasonKeys = mapReasonKeys(selectedReasonSignals, evaluations);
    const reasons = mapReasonCopy(reasonKeys);
    const roleKey = resolveRole(normalizedAnswers, finalOutcome);
    const outcomeCopy = RESULT_COPY_BY_OUTCOME[finalOutcome];

    return {
        score,
        maxScore: MAX_SCORE,
        baseOutcome,
        finalOutcome,
        label: outcomeCopy.label,
        headline: outcomeCopy.headline,
        intro: outcomeCopy.intro,
        reasonKeys,
        reasons,
        roleKey,
        roleCopy: ROLE_COPY_BY_KEY[roleKey],
        cta: CTA_BY_OUTCOME[finalOutcome],
        debug: {
            normalizedAnswers,
            triggeredGuardrails,
            signalScores: POSITIVE_REASON_PRIORITY.reduce<Record<AssessmentSignalKey, number>>((accumulator, signal) => {
                accumulator[signal] = evaluations[signal].score;
                return accumulator;
            }, {} as Record<AssessmentSignalKey, number>),
            signalReasonKeys: POSITIVE_REASON_PRIORITY.reduce<Record<AssessmentSignalKey, ReasonKey>>(
                (accumulator, signal) => {
                    accumulator[signal] = evaluations[signal].reasonKey;
                    return accumulator;
                },
                {} as Record<AssessmentSignalKey, ReasonKey>,
            ),
            positiveSignals: POSITIVE_REASON_PRIORITY.filter((signal) => evaluations[signal].isPositive),
            blockerSignals: BLOCKER_REASON_PRIORITY.filter((signal) => evaluations[signal].isBlocker),
            moderateSignals: POSITIVE_REASON_PRIORITY.filter((signal) => evaluations[signal].isModerate),
            selectedReasonSignals,
            matchedRole: roleKey,
        },
    };
}

File: backend/tests/pinterest_fit/engine.py
from __future__ import annotations

from tests.pinterest_fit.models import Answers, Evaluation, Outcome, SignalName

POSITIVE_PRIORITY = ("category", "offer", "assets", "website", "support", "ads", "goal")
BLOCKER_PRIORITY = ("website", "assets", "offer", "support", "ads", "category", "goal")
NOT_RIGHT_NOW_FALLBACK_PRIORITY = (
    "goal",
    "support",
    "ads",
    "category",
    "offer",
    "assets",
    "website",
)


def evaluate(answers: Answers) -> Evaluation:
    warnings: list[str] = []
    total_score = score_total(answers)
    base_outcome = classify_base_outcome(total_score)
    final_outcome, applied_guardrails = apply_guardrails(answers, base_outcome)
    reason_keys = select_reasons(answers, final_outcome, warnings)
    role_key = assign_role(answers, final_outcome, warnings)
    return Evaluation(
        total_score=total_score,
        base_outcome=base_outcome,
        final_outcome=final_outcome,
        role_key=role_key,
        reason_keys=reason_keys,
        applied_guardrails=applied_guardrails,
        warnings=tuple(warnings),
    )


def score_total(answers: Answers) -> int:
    return (
        answers.q1_category_fit
        + answers.q2_offer_proven
        + answers.q3_assets
        + answers.q4_website
        + answers.q5_goal_fit
        + answers.q6_support_readiness
        + answers.q7_ads_openness
    )


def classify_base_outcome(total_score: int) -> Outcome:
    if total_score >= 18:
        return "strong_fit"
    if total_score >= 10:
        return "possible_fit"
    return "not_right_now"


def apply_guardrails(answers: Answers, base_outcome: Outcome) -> tuple[Outcome, tuple[str, ...]]:
    outcome = base_outcome
    applied: list[str] = []

    if (
        answers.q5_goal_type == "sales"
        and answers.q6_support_readiness <= 1
        and answers.q7_ads_openness <= 1
        and outcome == "strong_fit"
    ):
        outcome = "possible_fit"
        applied.append("guardrail_a")

    if answers.q3_assets == 0 and answers.q4_website == 0:
        outcome = "not_right_now"
        applied.append("guardrail_b")

    if answers.q1_category_fit == 1 and answers.q2_offer_proven <= 1:
        outcome = "not_right_now"
        applied.append("guardrail_c")

    return outcome, tuple(applied)


def select_reasons(
    answers: Answers, final_outcome: Outcome, warnings: list[str]
) -> tuple[str, str, str]:
    if final_outcome == "strong_fit":
        category_reason = reason_key_for_signal(answers, "category")
        foundation_signal = pick_best_signal(
            answers,
            ("offer", "assets", "website"),
            tie_order=("offer", "assets", "website"),
        )
        readiness_signal = pick_best_signal(
            answers,
            ("support", "ads", "goal"),
            tie_order=("support", "ads", "goal"),
        )
        return (
            category_reason,
            reason_key_for_signal(answers, foundation_signal),
            reason_key_for_signal(answers, readiness_signal),
        )

    if final_outcome == "possible_fit":
        positives = [signal for signal in POSITIVE_PRIORITY if is_positive(answers, signal)]
        selected_signals: list[SignalName] = []
        if positives:
            selected_signals.append(positives[0])

        if len(positives) >= 2:
            selected_signals.append(positives[1])
        else:
            moderate = next((signal for signal in POSITIVE_PRIORITY if is_moderate(answers, signal)), None)
            if moderate is not None:
                selected_signals.append(moderate)
            else:
                fallback = next(
                    signal
                    for signal in POSITIVE_PRIORITY
                    if signal not in selected_signals
                )
                selected_signals.append(fallback)

        blockers = [
            signal
            for signal in BLOCKER_PRIORITY
            if is_blocker(answers, signal) and signal not in selected_signals
        ]
        if blockers:
            selected_signals.append(blockers[0])

        selected = [reason_key_for_signal(answers, signal) for signal in selected_signals]
        return ensure_three(selected, answers, warnings)

    blocker_reasons = [
        reason_key_for_signal(answers, signal)
        for signal in BLOCKER_PRIORITY
        if is_blocker(answers, signal)
    ]
    selected = blocker_reasons[:3]
    if len(selected) < 3:
        selected.extend(fill_not_right_now_reasons(answers, selected, warnings))
    return ensure_three(selected, answers, warnings)


def fill_not_right_now_reasons(
    answers: Answers, selected: list[str], warnings: list[str]
) -> list[str]:
    selected_signals = {signal_for_reason(reason) for reason in selected}
    remaining_signals = [signal for signal in POSITIVE_PRIORITY if signal not in selected_signals]
    filled: list[str] = []

    while len(selected) + len(filled) < 3:
        scores = {signal: signal_score(answers, signal) for signal in remaining_signals}
        next_score = min(scores.values())
        tied_signals = [signal for signal in remaining_signals if scores[signal] == next_score]
        tied_signals.sort(key=NOT_RIGHT_NOW_FALLBACK_PRIORITY.index)
        chosen = tied_signals[0]
        filled.append(reason_key_for_signal(answers, chosen))
        remaining_signals.remove(chosen)

    return filled


def ensure_three(selected: list[str], answers: Answers, warnings: list[str]) -> tuple[str, str, str]:
    deduped: list[str] = []
    for reason in selected:
        if reason not in deduped:
            deduped.append(reason)

    if len(deduped) < 3:
        warnings.append("reason_fill_required")
        for signal in POSITIVE_PRIORITY:
            reason = reason_key_for_signal(answers, signal)
            if reason not in deduped:
                deduped.append(reason)
            if len(deduped) == 3:
                break

    return deduped[0], deduped[1], deduped[2]


def assign_role(answers: Answers, final_outcome: Outcome, warnings: list[str]) -> str | None:
    role_key: str | None = None

    if final_outcome == "not_right_now":
        role_key = "not_priority_yet"
    elif (
        answers.q5_goal_type == "sales"
        and answers.q2_offer_proven >= 3
        and answers.q4_website >= 3
        and answers.q7_ads_openness >= 2
    ):
        role_key = "sales_with_ads_support"
    elif answers.q5_goal_type == "retargeting" and final_outcome != "not_right_now":
        role_key = "warm_audience_support"
    elif (
        answers.q1_category_fit >= 3
        and answers.q3_assets >= 3
        and answers.q4_website >= 3
        and answers.q5_goal_type in ("discovery", "traffic")
    ):
        role_key = "discovery_traffic"
    elif (
        answers.q1_category_fit <= 2
        and answers.q2_offer_proven >= 3
        and answers.q3_assets >= 3
        and answers.q4_website >= 3
        and answers.q5_goal_type in ("discovery", "traffic", "launches")
        and final_outcome != "not_right_now"
    ):
        role_key = "selective_test_channel"
    elif (
        answers.q1_category_fit >= 3
        and answers.q3_assets >= 1
        and answers.q4_website >= 1
        and answers.q5_goal_type in ("discovery", "traffic", "launches")
    ):
        role_key = "organic_first_ads_later"
    elif final_outcome != "not_right_now":
        role_key = "foundation_first"

    return role_key


def pick_best_signal(
    answers: Answers, candidates: tuple[SignalName, ...], tie_order: tuple[SignalName, ...]
) -> SignalName:
    best_score = max(signal_score(answers, signal) for signal in candidates)
    tied = [signal for signal in candidates if signal_score(answers, signal) == best_score]
    tied.sort(key=tie_order.index)
    return tied[0]


def is_positive(answers: Answers, signal: SignalName) -> bool:
    if signal == "category":
        return answers.q1_category_fit >= 3
    if signal == "offer":
        return answers.q2_offer_proven >= 3
    if signal == "assets":
        return answers.q3_assets >= 3
    if signal == "website":
        return answers.q4_website >= 3
    if signal == "support":
        return answers.q6_support_readiness >= 2
    if signal == "ads":
        return answers.q7_ads_openness >= 2
    return answers.q5_goal_type in ("discovery", "traffic", "launches", "retargeting")


def is_blocker(answers: Answers, signal: SignalName) -> bool:
    if signal == "category":
        return answers.q1_category_fit <= 2
    if signal == "offer":
        return answers.q2_offer_proven <= 1
    if signal == "assets":
        return answers.q3_assets <= 1
    if signal == "website":
        return answers.q4_website <= 1
    if signal == "support":
        return answers.q6_support_readiness <= 1
    if signal == "ads":
        return answers.q7_ads_openness <= 1
    return answers.q5_goal_type == "sales"


def is_moderate(answers: Answers, signal: SignalName) -> bool:
    return not is_positive(answers, signal) and not is_blocker(answers, signal)


def signal_score(answers: Answers, signal: SignalName) -> int:
    if signal == "category":
        return answers.q1_category_fit
    if signal == "offer":
        return answers.q2_offer_proven
    if signal == "assets":
        return answers.q3_assets
    if signal == "website":
        return answers.q4_website
    if signal == "support":
        return answers.q6_support_readiness
    if signal == "ads":
        return answers.q7_ads_openness
    return answers.q5_goal_fit


def reason_key_for_signal(answers: Answers, signal: SignalName) -> str:
    if signal == "category":
        return {
            4: "reason_category_strong",
            3: "reason_category_good",
            2: "reason_category_maybe",
            1: "reason_category_weak",
        }[answers.q1_category_fit]
    if signal == "offer":
        return {
            4: "reason_offer_proven",
            3: "reason_offer_some_traction",
            1: "reason_offer_early",
            0: "reason_offer_unproven",
        }[answers.q2_offer_proven]
    if signal == "assets":
        return {
            4: "reason_assets_strong",
            3: "reason_assets_decent",
            1: "reason_assets_limited",
            0: "reason_assets_weak",
        }[answers.q3_assets]
    if signal == "website":
        return {
            4: "reason_site_ready",
            3: "reason_site_solid",
            1: "reason_site_friction",
            0: "reason_site_not_ready",
        }[answers.q4_website]
    if signal == "support":
        return {
            3: "reason_support_ready",
            2: "reason_support_open",
            1: "reason_support_cautious",
            0: "reason_support_not_committed",
        }[answers.q6_support_readiness]
    if signal == "ads":
        return {
            3: "reason_ads_open",
            2: "reason_ads_later",
            1: "reason_ads_unsure",
            0: "reason_ads_not_open",
        }[answers.q7_ads_openness]
    return {
        "discovery": "reason_goal_discovery",
        "traffic": "reason_goal_traffic",
        "launches": "reason_goal_launches",
        "retargeting": "reason_goal_retargeting",
        "sales": "reason_goal_sales_caution",
    }[answers.q5_goal_type]


def signal_for_reason(reason_key: str) -> SignalName:
    if reason_key.startswith("reason_category_"):
        return "category"
    if reason_key.startswith("reason_offer_"):
        return "offer"
    if reason_key.startswith("reason_assets_"):
        return "assets"
    if reason_key.startswith("reason_site_"):
        return "website"
    if reason_key.startswith("reason_goal_"):
        return "goal"
    if reason_key.startswith("reason_support_"):
        return "support"
    return "ads"
